created method for swh api for getting content info

sanketvarpe opened the pull request:
### Created first method along with test suites and successful test screenshots.

#### Task completed:

- Created a method and its tests to get content information about any project using SWH api which requires hash type and hash value to fetch data

I have just created one method for now because it needs a review about coding style and manner. I have explored out coding style from docs and reading other codes and tried to follow it. 

I'll be happy to take suggestions and enhancement in context of coding style, testing style, and in any other view.

PFA screenshot of tests:
![Screenshot from 2022-01-31 12-24-09](https://user-images.githubusercontent.com/67329087/151750951-3d5c1dfe-1cc9-4ab0-985c-c4202770cf85.png)

@pombredanne and @TG1999 please review.

sanketvarpe committed:
created method for swh api for getting content info

sanketvarpe committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

created method for swh api for getting content info

Given updates on first PR completed.

Docstring placed inside function

Class created for returning custom reponse from api functions.

Updated given changes regarding custom class, naming and commenting conventions.

Concrete attributes added in CustomApiResponse class instead of generic

updated class definition and function comments

comments for class attributes,assertion for errors,status code checking added

Made changes in comments and argument passing style.

# This is the commit message #2:

Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
created method for swh api for getting content info

Given updates on first PR completed.

Docstring placed inside function

Class created for returning custom reponse from api functions.

Updated given changes regarding custom class, naming and commenting conventions.

Concrete attributes added in CustomApiResponse class instead of generic

updated class definition and function comments

comments for class attributes,assertion for errors,status code checking added

Made changes in comments and argument passing style.

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+
+base_content_swh_api_url = "https://archive.softwareheritage.org/api/1/content/"
+
+#this method is used to get information about content (aka a "blob") object. 
```
This should be a docstring

pombredanne replied:
And BTW this is a function, not a method

sanketvarpe replied:
Okay sure, i'll update it.

sanketvarpe replied:
`"""
this function returns information like checksums, data_url, file_url, license_url
based on hash_type and hash as a json object
"""

def get_content_information(hash_type, hash):
    api = f"{base_swh_api_url}content/{hash_type}:{hash}/"
    response = requests.get(api)
    return response.json()`

This is the updated one.

sanketvarpe replied:
`"""
this function returns information like checksums, data_url, file_url, license_url
based on hash_type and hash as a json object
"""`

This is updated one

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+#
+
+
+import requests as req
```
Why using a shortcut? readability matters a lot

sanketvarpe replied:
Okay sure, i'll update it.

sanketvarpe replied:
`import requests`
Updated one.

pombredanne commented on tests/contentmethods/test_methods.py:
```diff
@@ -0,0 +1,31 @@
+data_dir_path = os.path.join(os.path.dirname(__file__), "data")
+
+#input test data
+input_test_data = json.load(open(os.path.join(data_dir_path, "input.json")))
```
Why is this not in the test function?

sanketvarpe replied:
Because data would be stored in same file for current test method and all upcoming test methods in this test_methods.py. So instead of reading again and again for every test method why not read it once and use it in all methods. This is my logic any change suggestion please let me know.

sanketvarpe replied:
As you said i have removed data folder and instead added i/p and o/p data inside test functions only.

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+#
```
I am not sure contentmethods and methods are super useful names...... what are these about? give a name that hints at  what they do, not what they are.

sanketvarpe replied:
Actually want to discuss about it. Because SWH api docs have api's nearly in 3-4 for subtopics like about content, about origin, about ops(release, builds etc) then some miscllaneous topics. So i thought we'll make seorate folders for all these seprate topics and will keep their related methods/api in one file inside those folders. So want discuss how to go about it.

sanketvarpe replied:
Actuall tushar suggested me to just keep one file in src folder so i did that way and new file name is swh_api_functions.py

pombredanne commented on requirements-dev.txt:
```diff
@@ -0,0 +1 @@
+requests>=2.27
```
No need to duplicate this for dev

sanketvarpe replied:
Okay sure, i'll update it.

sanketvarpe replied:
Updated.

pombredanne commented on CHANGELOG.rst:
```diff
@@ -6,3 +6,13 @@ v0.0.0
 
 *xxxx-xx-xx* -- Initial release.
+
+Methods/API Update
```
A changelog is there to list big important changes. Tests are rarely mentioned in this

sanketvarpe replied:
Okay so should i not even mention about new methods being added. Or where should i maintain docs for these methods??

sanketvarpe replied:
Updated.

pombredanne commented on tests/contentmethods/test_methods.py:
```diff
@@ -0,0 +1,31 @@
+    #json needs to do dumps to equate two jsons
+    expected_output = json.dumps(expected_output, sort_keys=True)
+    
+    assert json.dumps(methods.get_content_information(hash_type, hash), sort_keys=True) == expected_output
```
Is this doing live HTTP requests? If yes, use a mock instead

pombredanne replied:
And do not sort keys... also why not having at least this simple tests not using data files but rather using the data directly in the test function for clarity?

sanketvarpe replied:
Ok will look into it.

pombredanne commented on tests/contentmethods/data/expected.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "output_get_content_information": {
```
what does this name mean? that as cryptic as it gets. It should be descriptive... what does this mapping contain?

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+#this method is used to get information about content (aka a "blob") object. 
+#In the archive, a content object is identified based on checksum values computed using various hashing algorithms.
+
+def get_content_information(hash_type, hash):
```
What would be a better name for this?

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+
+def get_content_information(hash_type, hash):
+    endpoint = f"{hash_type}:{hash}/"
+    api = base_content_swh_api_url + endpoint
```
Why using a + here when using an fstring above?

pombredanne commented on src/contentmethods/methods.py:
```diff
@@ -0,0 +1,27 @@
+    res = req.get(api)
+
+    if res.status_code == 404:
+        return {"msg":"No data found"}
```
what does msg mean?

sanketvarpe commented:
@pombredanne Almost all of the above updates have been completed. So should i push them again to review now or should i implement other functions and then push them all together for review

sanketvarpe committed:
Given updates on first PR completed.

TG1999 commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,23 @@
+base_swh_api_url = "https://archive.softwareheritage.org/api/1/"
+
+"""
+this function returns information like checksums, data_url, file_url, license_url
```
add docstring inside function definition

sanketvarpe replied:
Okay

sanketvarpe committed:
Docstring placed inside function

TG1999 commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,23 @@
+    """
+    api = f"{base_swh_api_url}content/{hash_type}:{hash}/"
+    response = requests.get(api)
+    return response.json()
```
Instead of returning a json we should return some type of object which we can obtain by mapping json with some kind of class object, since it returns `checksums`, `data_url`, `file_url`, `license_url`. We should create a class with at least these as attributes for now.

sanketvarpe replied:
Okay so you mean we must have class where it'll have attributes which we get in response from SWH api and whenever we get reponse we'll map those keys from json to our class and make new obj fromt that class with these attributes and return that object right??

sanketvarpe replied:
How that class object will benifit our end user when we return it.And how it would be diffrent from an json obj in view of usage?? Even if its a class obj will access out o/p like obj.key right??Will you elaborate a little about its adv and usage plz??

TG1999 replied:
Yes

TG1999 replied:
You are just returning a json object that you are getting from the API. We can not expect every time the required data to be there in JSON response, it may or may not be there.

sanketvarpe replied:
> You are just returning a json object that you are getting from the API. We can not expect every time data to be there in JSON, it may or may not be there.

I went through their docs and it says as follows:
**API access is over HTTPS.
All API endpoints are rooted at https://archive.softwareheritage.org/api/1/.
Data is sent and received as JSON by default.**

sanketvarpe replied:
> You are just returning a json object that you are getting from the API. We can not expect every time data to be there in JSON, it may or may not be there.

I went through their docs and it says as follows:
**API access is over HTTPS.
All API endpoints are rooted at https://archive.softwareheritage.org/api/1/.
Data is sent and received as JSON by default.**

So i guess format is json for all api's still if we want our own format we can. Just want make clear that every api response will have diffrent attributes so we'll need to accomodate all of'em. So instead of that can't we have a simple json data structure like
`{
data: 'data from api',
status_code: 'status_code',
etc...
}`
This is just my suggestion still if we want class method i can do that.
Or we can have class and in that class we can a property/variable which will be json obj with some keys and then we can update api response json in that obj and in future we can add whatever we want in that class??

TG1999 replied:
I didn't meant that the data will not be in JSON format, I meant that the required data or the attributes you want may or may not be in the JSON

sanketvarpe replied:
Sorry but still not getting you.for now i'll create a class with suppose data_url, checksum etc as attributes and when i'll get response form api i'll make new obj of that class and set those class attributes from the json which i'll get from api in reponse.Then you can review and suggest.Ok??

TG1999 replied:
Yeah please make a class object.

sanketvarpe replied:
> Yeah please make a class object.

Implemented class and used it in functions. Also Completed updates given by you.
Please review.
Thanks

sanketvarpe replied:
@pombredanne will you please look into this thread.
Thanks.

TG1999 commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,23 @@
+    """
+    api = f"{base_swh_api_url}content/{hash_type}:{hash}/"
+    response = requests.get(api)
+    return response.json()
```
Use Black for formatting

sanketvarpe replied:
Ok

TG1999 commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,23 @@
+
+import requests
+
+base_swh_api_url = "https://archive.softwareheritage.org/api/1/"
```
This will be a constant, so I think we should write it as `BASE_SWH_API_URL`

sanketvarpe replied:
Okay

sanketvarpe committed:
Class created for returning custom reponse from api functions.

pombredanne commented on src/custom_api_response.py:
```diff
@@ -0,0 +1,32 @@
+        for key in data:
+            self.key = data[key]
+
+    """
```
A docstring should right after a function definition, also you may want to check the spelling: obj? claa? 
But here, the `__eq__` dunder method being a special well known method, there is really no need a docstring at all ... so you are better off not having one entirely

pombredanne commented on src/custom_api_response.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class CustomApiResponse:
+    """
```
Why would you need such a class?  And why split your code in two python files? One file is simpler and easier to read in this case. Python is not like Java. We like things simple.
Try running this:
```
$ python
>>> import this
```

sanketvarpe replied:
> Why would you need such a class? And why split your code in two python files? One file is simpler and easier to read in this case. Python is not like Java. We like things simple. Try running this:
> 
> ```
> $ python
> >>> import this
> ```

Actually @TG1999 suggested me to have our own class and then while returning response from a function make new object of this class and set those json attributes as this class attribute.
But actually i also didn't got why is it needed if we are getting json in response from SWH api which is standard right??

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+
+import requests
+
+# class to genrate our own custom response
```
Having a single script will remove imports

sanketvarpe replied:
Yeah ok.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+# class to genrate our own custom response
+from custom_api_response import CustomApiResponse
+
+BASE_SWH_API_URL = "https://archive.softwareheritage.org/api/1/"
```
Will this be ever used outside of the function below?

sanketvarpe replied:
If by this if you mean BASE_SWH_API_URL, no it wont be used outside of any function.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def get_content_information(hash_type, hash):
+
```
Remove this empty line

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+def get_content_information(hash_type, hash):
+
+    """
+    this function returns information like checksums, data_url, file_url, license_url
```
Check the style of comments used in scancode-toolkit code
We like imperative style for docstrings. This is eventually doc that's designed to end up in generated HTML web pages on readthedocs

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+BASE_SWH_API_URL = "https://archive.softwareheritage.org/api/1/"
+
+
+def get_content_information(hash_type, hash):
```
What about using `checksum_type` and `checksum` as names instead?

sanketvarpe replied:
Yeah seems better.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,29 @@
+    api = f"{BASE_SWH_API_URL}content/{hash_type}:{hash}/"
+    response = requests.get(api)
+    response = response.json()
+    custom_response = CustomApiResponse(response)
```
I would rather see a model object that contains actual concrete fields than a generic data class like this. The value this adds is what when compared with a plain dictionary?
Think about which value this library brings. I think it can be more than merely making a request call?

sanketvarpe replied:
@pombredanne  Could you please elaborate a little how should we go about returning response from functions after getting it from an api??.
I am little confused there. Like why can't we return json which we get from an api?? If can't what extra feature we need other than json data. I am stucked there itself for a while now. I am not clear about requirement.

Thanks.

pombredanne replied:
We want structured data in a class that has concrete named instance attributes and not some generic object that can have any attributes. You want a real schema of sorts. Please look at crafting a class https://docs.python.org/3/tutorial/classes.html

pombredanne commented on tests/test_swh_api_functions.py:
```diff
@@ -0,0 +1,48 @@
+    hash = "fe95a46679d128ff167b7c55df5d02356c5a1ae1"
+    # output data
+    json_output = {
+        "checksums": {
```
All these should be the fields **defined** as attributes on an object not just attached to it. May be using a simple attrs or dataclass for a start would go a long way

sanketvarpe replied:
Sorry but i didn't get it??. Do you mean like i shoud have an object like input_obj and then set these attributes to it??
Like this `{checksum_type:' ', checksum: ' '}` ??

pombredanne replied:
@sanketvarpe I am only restating exactly the same thing that @TG1999 said earlier in https://github.com/nexB/heritedcode/pull/1#discussion_r796337937 with different words: 

> Instead of returning a json we should return some type of object which we can obtain by mapping json with some kind of class object, since it returns checksums, data_url, file_url, license_url. We should create a class with at least these as attributes for now.

pombredanne reviewed (changes requested):
Thanks!
See some comments for your consideration.

pombredanne commented on src/custom_api_response.py:
```diff
@@ -0,0 +1,32 @@
+    """
+
+    def __init__(self, data):
+        for key in data:
```
If you need to access all items, use instead `for key, value in data.items():`

pombredanne commented on tests/test_swh_api_functions.py:
```diff
@@ -0,0 +1,48 @@
+    api = f"{BASE_SWH_API_URL}content/{hash_type}:{hash}/"
+    requests_mock.get(api, json=json_output)
+
+    assert (
```
you should test the value of each attributes instead. Your test tests nothing as of now.

sanketvarpe replied:
They will be compared in class where i have  __eq__ method which i have corrected now.

pombredanne replied:
You still need a proper class with concrete attributes.

sanketvarpe replied:
> You still need a proper class with concrete attributes.

@pombredanne Yeah got that, working on it and will push it soon.

pombredanne replied:
:+1:

pombredanne commented on src/custom_api_response.py:
```diff
@@ -0,0 +1,32 @@
+
+    def __init__(self, data):
+        for key in data:
+            self.key = data[key]
```
BTW, I am pretty this does not what you expect it to do:
```Python
from custom_api_response import CustomApiResponse
input = {
  "data_url": "https://archive.softwareheritage.org/api/1/content/sha1_git:fe95a46679d128ff167b7c55df5d02356c5a1ae1/raw/",
  "filetype_url": "https://archive.softwareheritage.org/api/1/content/sha1_git:fe95a46679d128ff167b7c55df5d02356c5a1ae1/filetype/",
  "language_url": "https://archive.softwareheritage.org/api/1/content/sha1_git:fe95a46679d128ff167b7c55df5d02356c5a1ae1/language/",
  "length": 151810,
  "license_url": "https://archive.softwareheritage.org/api/1/content/sha1_git:fe95a46679d128ff167b7c55df5d02356c5a1ae1/license/",
  "status": "visible",
}

car = CustomApiResponse(input)

car.data_url
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'CustomApiResponse' object has no attribute 'data_url'

car.key
'visible'
```

so this is not right at all

sanketvarpe replied:
Yeah got that, my bad.
Thanks.

sanketvarpe commented:
> Thanks! See some comments for your consideration.

Thanks for all your reviews. :)

sanketvarpe committed:
Updated given changes regarding custom class, naming and commenting conventions.

sanketvarpe commented:
@pombredanne Pushed updated code. https://github.com/nexB/heritedcode/pull/1/commits/ab68cc1cebf1b2c53c361fb30fc4fdf05b0acb40

- Moved all code in single file.
- Updated variable namings.
- Updated commenting style
- Corrected CustomApiResponse dataclass parameter defining bug in class constructor.
- Now in test `__eq__`  method of class CustomApiResponse is checking each and every attribute within itself.
As we will be having multiple diffrent attributes from SWH api, i have kept our class constructor dynamic to accomodate all attributes.
Please review.
Thank You.

sanketvarpe committed:
Concrete attributes added in CustomApiResponse class instead of generic

sanketvarpe commented:
@pombredanne Added concrete attributes in our CustomApiResponse class instead of generic attributes. https://github.com/nexB/heritedcode/pull/1/commits/723895c0cfe7a0f3b7685f921c0e509e0af97e6b
Please review it thanks.

sanketvarpe commented:
> @pombredanne Added concrete attributes in our CustomApiResponse class instead of generic attributes. [723895c](https://github.com/nexB/heritedcode/commit/723895c0cfe7a0f3b7685f921c0e509e0af97e6b) Please review it thanks.

@pombredanne Have you reviewed these new changes??Should i implement rest of the API's if these changes are ok??

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+
+class CustomApiResponse:
+    """
+    this class represents response data structure for returning response
```
I am sure we can find a better description... Ask around to someone to read this and try to make sense of it... it tell me nothing about the object,  and `"CustomApiResponse"  is also something to improve.
Let's make this perfect ;)

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+            if key in data:
+                setattr(self, key, data[key])
+
+    def __eq__(self, second_class_obj):
```
I am still not sure why you need this.... can you elaborate?

sanketvarpe replied:
@pombredanne  If you are asking about __eq__, its required when i compare two objects while testing so that it compares each and every attribute of two objects.
And if you are asking about `for key in vars(self)` this loop, its required to set attributes as of now becuase there are lot of attributes so its difficult to pass each of'em in contructor and set them bcoz of sequence and all.
Thanks.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+    """
+
+    def __init__(self, data):
+        self.checksums = None
```
Can you add a short comment above each item that describes what this is? Als you should pass each attribute as an argument and use a @classmethod "from_swh" methoo that creates an instance from the SWH data instead...

sanketvarpe replied:
@pombredanne I have a question here, as we will be having lot of attributes from swh and some of them will be present in some api's response and some won't be. But as you said we need to pass each and every attribute as an arg, so if we dont have any attribute in some api response thats not an error right?? we'll need to pass None int that case.
So basically from my point of view, passing each attribute as an arg seems little tedious instead of a dict/json. Every time i pass arg i'll have to check its sequence in constructor, and even if any arg is missing from total args in class that does not mean its an error because every api call will have diff set of attributes in response not all of'em will be present every time.

Whats your take on this??Hope you got me??
Thanks!! :)

pombredanne replied:
> But as you said we need to pass each and every attribute as an arg, so if we dont have any attribute in some api response thats not an error right??

That's an error alright. Or that's just return None, but that's not something that can be ignored silently, 

> So basically from my point of view, passing each attribute as an arg seems little tedious instead of a dict/json. 

Tedious but that's what is correct... we have  concrete data with a schema and the schema exists always  :) ... that's also why there are utilities like https://www.attrs.org that remove a lot of the gruntwork of the boilerplate.
In a way to me, applying some schemata to some external data is part of the essence of programming: you apply order to chaos and apply intelligence and structure to understand the external world. 

To may be better understand why you need concrete attributes vs. a plain schemaless dict, please view this https://martinfowler.com/tags/noSQL.html and https://martinfowler.com/articles/schemaless/ and this video http://www.youtube.com/watch?v=qI_g07C_Q5I


>  Every time i pass arg i'll have to check its sequence in constructor, and even if any arg is missing from total args in class that does not mean its an error because every api call will have diff set of attributes in response not all of'em will be present every time.

That's why you need a factory that will pick the field you support.

sanketvarpe replied:
Okay..Got it..
Thanks.. :)

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+        self.license_url = None
+        self.status = None
+
+        for key in vars(self):
```
Create a @classmethod "from_swh" method instead that creates an instance from the SWH data instead... 
Also use concrete attribute names and not some possibly random iteration. Your code would always work as it is ... and would not fail if it has more or missing attributes. You need it to fail in these cases.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+
+def get_content_information(checksum_type, checksum):
+    """
+    Makes call to Software Heritage api to fetch information like checksum, data_url
```
Use an imperative style for comment typically starting with `Return`
```
"""
Return ... 
"""
```

Think about you reading this comment two years from now: will it be something that will help you grok what this does?

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+    Makes call to Software Heritage api to fetch information like checksum, data_url
+    file_url etc.
+    """
+    api = f"{BASE_SWH_API_URL}content/{checksum_type}:{checksum}/"
```
What if `checksum_type = None` and `checksum= [1,2,3]`? You should validated the data you receive and error out otherwise.

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+    api = f"{BASE_SWH_API_URL}content/{checksum_type}:{checksum}/"
+    response = requests.get(api)
+    response = response.json()
+    custom_response = CustomApiResponse(response)
```
Why `custom_response` is this really what you get? What is this about? Can you find a better name?

pombredanne commented on tests/test_swh_api_functions.py:
```diff
@@ -0,0 +1,44 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# heritedcode is a trademark of nexB Inc.
```
```suggestion
```

pombredanne commented on src/swh_api_functions.py:
```diff
@@ -0,0 +1,50 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# heritedcode is a trademark of nexB Inc.
```
```suggestion
```

pombredanne commented on tests/test_swh_api_functions.py:
```diff
@@ -0,0 +1,44 @@
+        "license_url": "https://archive.softwareheritage.org/api/1/content/sha1_git:fe95a46679d128ff167b7c55df5d02356c5a1ae1/license/",
+        "status": "visible",
+    }
+    custom_expected_output = swh_api_functions.CustomApiResponse(json_output)
```
Why this custom prefix?

pombredanne replied:
And why ` swh_api_functions`? why not  `heritedcode` instead?

pombredanne reviewed (changes requested):
Thanks! much improved... See my comments inline for next steps.

sanketvarpe commented:
> Thanks! much improved... See my comments inline for next steps.

Thanks for your comments! Will definitely look into those and improve on them. :)

sanketvarpe committed:
updated class definition and function comments

sanketvarpe commented:
@pombredanne Updated class definition,now passing each and every attributes to class constructor and added @classmethod. Also updated on previous reviews.  https://github.com/nexB/heritedcode/commit/cb275575cdf485cb8a2f4d93a4ba279a29ef7e6b
Please review.

TG1999 commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,88 @@
+        raise Exception("checksum_type and checksum must be string")
+
+    api = f"{BASE_SWH_API_URL}content/{checksum_type}:{checksum}/"
+    response = requests.get(api)
```
Please check status of response

TG1999 commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    Returns a class object with attributes like checksum, data_url, file_url etc. based on checksum_type and checksum.
+    """
+    if type(checksum_type) != str or type(checksum) != str:
```
Please use assertion here

TG1999 commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,88 @@
+    Returns an instance of ApiResponse class.
+    """
+
+    @classmethod
```
How is it different from init function ?

sanketvarpe replied:
Actually @pombredanne asked to keep a @classmethod and genrate instance of class from there only.

TG1999 commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,88 @@
+        license_url,
+        status,
+    ):
+        self.checksums = checksums
```
Please add some docstring

TG1999 reviewed (commented):
We are very close in closing this PR, please check these comments

sanketvarpe committed:
comments for class attributes,assertion for errors,status code checking added

sanketvarpe commented:
> We are very close in closing this PR, please check these comments

@TG1999 @pombredanne Pushed updated code. https://github.com/nexB/heritedcode/pull/1/commits/ea01376d4f0a18017d3f6e937ac2793d32b20133
Please review it.
Thanks.

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+
+class ApiResponse:
+    """
+    this class represents response data structure for returning response
```
```suggestion
    This class represents the SWH API response data structure
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+        license_url,
+        status,
+    ):
+        """represents diffrent type of checksums like sha1,sha1_git and their values"""
```
```suggestion
        """
        Initialize a new ApiResponse from a subset of interesting arguments.
        """
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+    """
+    assert (
+        type(checksum_type) == str and type(checksum) == str
+    ), "checksum_type and checksum must be string"
```
When you have error messages, it is always best to have enough details to understand the error... otherwise it is hard to fix the issue. Something more verbose such as this may help:
 
```suggestion
    ), f"checksum_type: {checksum_type!r} and {checksum!r} must be strings, not {type(checksum_type)!r} and {type(checksum)!}"
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+        status,
+    ):
+        return cls(
+            checksums, data_url, filetype_url, language_url, length, license_url, status
```
Always prefer using named keyword arguments over positional arguments: this is more robust and avoid problems if arguments are renamed and moved in the future:

```suggestion
            checksums=checksums,
            data_url=data_url,
            filetype_url=filetype_url,
            language_url=language_url,
            length=length,
            license_url=license_url,
            status=status,
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+        return False
+
+    """
+    Returns an instance of ApiResponse class.
```
Use imperative style for your docstrings (and commit messages)... but also move this down as a proper doctrsing of `from_swh`
```suggestion
    Return an instance of ApiResponse class.
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+    ):
+        """represents diffrent type of checksums like sha1,sha1_git and their values"""
+        self.checksums = checksums
+        """represents data url for available raw data of any code"""
```
For these, you should prefer using comments... same for the other ones below. 

Also is this the URL to get the actual file content? If yes it may be best to reuse and "quote" the actual description provided by SWH or something very close to it.

And adding a sample (real) URL  helps a lot understanding what this is about. 

Here is a suggestion:
```suggestion
        # URL to download the content raw bytes. 
        # For example: https://archive.softwareheritage.org/api/1/content/sha1:0001c9f21990e5c61dd92f6a4a93d306987a1a31/raw/
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+
+def get_content_information(checksum_type, checksum):
+    """
+    Returns a class object with attributes like checksum, data_url, file_url etc. based on checksum_type and checksum.
```
Please use imperative style. Also 
```suggestion
    Return an ApiResponse object given a ``checksum_type`` and a ``cheksum`` value.
```

pombredanne commented on src/heritedcode_api_functions.py:
```diff
@@ -0,0 +1,98 @@
+    assert status_code != 400, "an invalid hash_type or hash has been provided"
+    assert status_code != 404, "requested content can not be found in the archive"
+    response_object = ApiResponse.from_swh(
+        response.get("checksums"),
```
As above, please always prefer using named, keyword arguments over positional ones. 
Also your `response` is a mapping. You expect certain keys to be there, so IMHO you should not "get" but directly obtain the item:

 ```suggestion
        checksums=response["checksums"],
```
... which would fail if checksums is not part of the response... but that's a feature and wanted: this way the code will stop working if the API changes at SWH and this is a good thing.

pombredanne reviewed (changes requested):
Thank you: you have made excellent updates. :+1: 
I have added a few extra review for your kind consideration!

sanketvarpe committed:
Made changes in comments and argument passing style.

sanketvarpe commented:
@pombredanne Made suggested changes and improvisations.
Please have a look and review it. https://github.com/nexB/heritedcode/commit/0cc7cfcb22fbcf166ecf0d07c7f10b869f254488
Thank You :)

pombredanne reviewed (changes requested):
Looking great!
a few things to wrap this up:
- The tests fails in the CI because you do not have "requests" yet in the setup.cfg `install_requires`
- You should squash your commit in a single commit with a proper message and DCO signoff

(I also enabled the DCO bot for signoff and Azure pipelines on the repo)

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Merge branch 'develop' of https://github.com/sanketvarpe/heritedcode into develop

sanketvarpe commented:
> Looking great! a few things to wrap this up:
> 
> * The tests fails in the CI because you do not have "requests" yet in the setup.cfg `install_requires`
> * You should squash your commit in a single commit with a proper message and DCO signoff
> 
> (I also enabled the DCO bot for signoff and Azure pipelines on the repo)

Hii @pombredanne I actually messed up while squashing because i am little new to this rebasing stuff..So created a seprate pull request. But didn't got time to work for past 4 days due to exams. Sorry for the mess. Back again, but what i see i have added requests and requests_mock in install_requires but still checks re failing and showing this https://dev.azure.com/nexB/heritedcode/_build/results?buildId=6074&view=logs&j=7af757f4-125b-5c1d-8ae5-4f59cbfd1686&t=b148fc87-f58e-57e0-9cb6-ca7d31e4c3e3&l=29. Tried to debug it but not able to get hold of it clearly will you please help me out here.Rest i learned about rebasing will complete that once this check gets passed on PR#1 only.
Thanks. :)

pombredanne commented:
> I actually messed up while squashing because i am little new to this rebasing stuff..So created a seprate pull request. But didn't got time to work for past 4 days due to exams. Sorry for the mess. Back again, but what i see i have added requests and requests_mock in install_requires but still checks re failing and showing this https://dev.azure.com/nexB/heritedcode/_build/results?buildId=6074&view=logs&j=7af757f4-125b-5c1d-8ae5-4f59cbfd1686&t=b148fc87-f58e-57e0-9cb6-ca7d31e4c3e3&l=29. Tried to debug it but not able to get hold of it clearly will you please help me out here.Rest i learned about rebasing will complete that once this check gets passed on PR#1 only.
Thanks. :)


That's OK, no worries. Learning git is part of the important things.
But the failure is not your making... let me check this over the week end!

sanketvarpe commented:
> > I actually messed up while squashing because i am little new to this rebasing stuff..So created a seprate pull request. But didn't got time to work for past 4 days due to exams. Sorry for the mess. Back again, but what i see i have added requests and requests_mock in install_requires but still checks re failing and showing this https://dev.azure.com/nexB/heritedcode/_build/results?buildId=6074&view=logs&j=7af757f4-125b-5c1d-8ae5-4f59cbfd1686&t=b148fc87-f58e-57e0-9cb6-ca7d31e4c3e3&l=29. Tried to debug it but not able to get hold of it clearly will you please help me out here.Rest i learned about rebasing will complete that once this check gets passed on PR#1 only.
> > Thanks. :)
> 
> That's OK, no worries. Learning git is part of the important things. But the failure is not your making... let me check this over the week end!

Sure. 
Thanks. :)

sanketvarpe commented:
> > > I actually messed up while squashing because i am little new to this rebasing stuff..So created a seprate pull request. But didn't got time to work for past 4 days due to exams. Sorry for the mess. Back again, but what i see i have added requests and requests_mock in install_requires but still checks re failing and showing this https://dev.azure.com/nexB/heritedcode/_build/results?buildId=6074&view=logs&j=7af757f4-125b-5c1d-8ae5-4f59cbfd1686&t=b148fc87-f58e-57e0-9cb6-ca7d31e4c3e3&l=29. Tried to debug it but not able to get hold of it clearly will you please help me out here.Rest i learned about rebasing will complete that once this check gets passed on PR#1 only.
> > > Thanks. :)
> > 
> > 
> > That's OK, no worries. Learning git is part of the important things. But the failure is not your making... let me check this over the week end!
> 
> Sure. Thanks. :)

@pombredanne Have you checked on this yet??
Thank You.

pombredanne reviewed (approved):
It can be refined alright and this is LGTM! 
Thank you. Merging!

pombredanne commented:
@sanketvarpe We are still missing the DCO on some commits though: can you confirm that you DCO "Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>" applies to the whole PR?

sanketvarpe commented:
> @sanketvarpe We are still missing the DCO on some commits though: can you confirm that you DCO "Signed-off-by: Sanket Varpe [varpesanket2000@gmail.com](mailto:varpesanket2000@gmail.com)" applies to the whole PR?

@pombredanne Sure i'll do that and let you know.
Thank You :)

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
# This is a combination of 5 commits.
# This is the 1st commit message:

Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

# This is the commit message #2:

Docstring placed inside function

# This is the commit message #3:

Class created for returning custom reponse from api functions.

# This is the commit message #4:

Updated given changes regarding custom class, naming and commenting conventions.

# This is the commit message #5:

Concrete attributes added in CustomApiResponse class instead of generic

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Merge branch 'develop' of https://github.com/sanketvarpe/heritedcode into develop

sanketvarpe committed:
created method for swh api for getting content info

sanketvarpe committed:
# This is a combination of 5 commits.
# This is the 1st commit message:

Given updates on first PR completed.

# This is the commit message #2:

Docstring placed inside function

# This is the commit message #3:

Class created for returning custom reponse from api functions.

# This is the commit message #4:

Updated given changes regarding custom class, naming and commenting conventions.

# This is the commit message #5:

Concrete attributes added in CustomApiResponse class instead of generic

sanketvarpe committed:
updated class definition and function comments

sanketvarpe committed:
Added functions to get basic content information from SWH API

Signed-off-by: Sanket Varpe <varpesanket2000@gmail.com>

sanketvarpe committed:
Merge branch 'develop' of https://github.com/sanketvarpe/heritedcode into develop

sanketvarpe commented:
@pombredanne Hello,was busy in college exams.Back now will start working on this PR.Need add DCO sign-off to all commits so that we can move ahead.But to add DCO sign-off need squash commits into one but i am kinda failing while squashing it up..commits are getting increased instaed of getting reduced. Will you please help me out here at initial?? Or else i have got another branch which i think contains 3 4 commits,i can try on that to get DCO sign off and make a brand new PR from that branch?? 
Thanks.

pombredanne commented:
> Will you please help me out here at initial?? Or else i have got another branch which i think contains 3 4 commits,i can try on that to get DCO sign off and make a brand new PR from that branch??

Sometimes this is indeed simpler to reset to create a clean commit stream and then force push the branch to get a cleaner stream.


https://dev.to/render/git-organized-a-better-git-flow-56go provides a decent approach.
You can also start a new branch if your prefer, but with the reset approach above that's not needed and you can ensure all your commits have a proper DCO alright!
